### ISSUE TITLE
Details abort Training

### ISSUE THREAD

**Issue by Dodojordi** [niki-amini-naieni/CountGD#32]
I want to know why it is necessary to construct a caption consisting of positive and negative labels for each photo during training, instead of directly inputting positive labels

**Comment by niki-amini-naieni** [comment]
Hi, thanks for the interest! We do this following the training procedure of Grounding DINO (see figure below). The negative labels allow us to train with a contrastive loss function so text and exemplar embeddings associated with a particular object (from the positive labels) are pushed closer to it while text not associated with a particular object (from the negative labels) are pushed away from it. This is done so that at inference, queries associated with a particular exemplar and/or text description can be matched and those with a low matching score can be filtered out.

<img width="487" alt="Image" src="https://github.com/user-attachments/assets/6dc7deb3-0f1a-444c-a04f-0ceff97a1b0b" />

**Comment by Dodojordi** [comment]
Hi, we have trained the code you provided and found that the results are somewhat different from those presented in the paper. Is there any possible reason ？


![Image](https://github.com/user-attachments/assets/5cabe67a-5b29-4f23-8018-b4c50b305033)

**Comment by niki-amini-naieni** [comment]
Hi, have you applied the adaptive cropping scheme or is this just without the cropping? Also please see: https://github.com/niki-amini-naieni/CountGD/blob/main/training.md.

**Comment by Dodojordi** [comment]
Hi，this is the test result of our training on the test set

![Image](https://github.com/user-attachments/assets/3730bd74-0d7b-4aed-893a-ae8982551cfb)

[output.txt](https://github.com/user-attachments/files/18559937/output.txt)

![Image](https://github.com/user-attachments/assets/aeb2f120-20cf-46f9-9617-1bf97bb2e938)

**Comment by niki-amini-naieni** [comment]
Hi, just to clarify, you mean training on the *training* set and testing on the *test* set, right? I think this is the issue I was mentioning about the RMSE being difficult to reproduce as documented [here](https://github.com/niki-amini-naieni/CountGD/blob/main/training.md) where we note: "We notice that due to examples in FSC-147 with high (> 900) objects and the sensitivity of the RMSE to large outlier errors from images with high counts, the RMSE obtained after training can vary significantly between runs. "

**Comment by QMiao-cs** [comment]
> Hi, just to clarify, you mean training on the _training_ set and testing on the _test_ set, right? I think this is the issue I was mentioning about the RMSE being difficult to reproduce as documented [here](https://github.com/niki-amini-naieni/CountGD/blob/main/training.md) where we note: "We notice that due to examples in FSC-147 with high (> 900) objects and the sensitivity of the RMSE to large outlier errors from images with high counts, the RMSE obtained after training can vary significantly between runs. "

Hello, it's great to see such interesting work. Since I’ve just started working on the counting task, I recently tried to replicate some of the works from 2024, but I couldn’t achieve the results reported in the papers. I wanted to ask, in this task, do you also use the traditional approach of training on the training set and testing on the test set? Or are there any other techniques involved?

**Comment by niki-amini-naieni** [comment]
You train on the training set and test on the test set. One can observe high variances in performance (especially in the RMSE) due to the imbalance of object counts in FSC-147 and the variations across different runs (due to factors such as the seed, hardware setup, differing libraries, or additional stochasticity in the algorithm).

**Comment by Eamon-0v0** [comment]
> If the cropping is the issue, maybe creating a softer cropping threshold will alleviate the issue. Try changing [this line](https://github.com/niki-amini-naieni/CountGD/blob/948399ca2eb26ce3faf8f4efcbab7353c952e273/engine_inference.py#L407): `if args.crop and pred_cnt == args.num_select` to `if args.crop and pred_cnt >= args.num_select - 15` If the RMSE is fixed by this, then the issue documented [here](https://github.com/niki-amini-naieni/CountGD/blob/main/training.md) was the problem.

Thanks! It‘s very useful, a strong 900 threshold does extremely amplify the error from randomness。

By adjusting the threshold of the crop, I achieved the same **Visual Exemplars & Text** results as reported in the paper！！！！😀

For the **Text only** experiment, are **TT-norm** and **crop** used as well?  Without them I still can't reproduce the results of the **Text only** experiment, but using them would mean that the **Visual Exemplars** information is needed.

**Comment by niki-amini-naieni** [comment]
This is great news! For the text only experiment, the TT-norm is not used. For the crop, we apply a simple cropping method that **does not require visual exemplars**. $CountGD_{txt}$ is a different model trained on just text than multi-modal CountGD trained on both exemplars and text as specified in the table caption (pasted below for reference).

<img width="836" alt="Image" src="https://github.com/user-attachments/assets/d7387453-38b0-466b-b081-b70549733e38" />
I can release the text only version once I have some time soon. We also find that multimodal CountGD performs fairly well given text only. It probably will not do as well as the model trained on only text, but try the commands below and see what performance you get. Note:  `--simple_crop` divides the image up into 4 pieces and just sums the counts in all the crops without relying on exemplars at any stage. I have not had time to test the below commands yet, so just let me know if you run into any bugs and I will get back to this ASAP.


`python -u main_inference.py --output_dir ./countgd_val -c config/cfg_fsc147_val.py --eval --datasets config/datasets_fsc147_val.json --pretrain_model_path checkpoints/checkpoint_fsc147_best.pth --options text_encoder_type=checkpoints/bert-base-uncased --simple_crop --num_exemplars 0`
 
`python -u main_inference.py --output_dir ./countgd_test -c config/cfg_fsc147_test.py --eval --datasets config/datasets_fsc147_test.json --pretrain_model_path checkpoints/checkpoint_fsc147_best.pth --options text_encoder_type=checkpoints/bert-base-uncased --simple_crop --num_exemplars 0`

**Comment by Eamon-0v0** [comment]
> 这是个好消息！对于纯文本实验，不使用 TT 范数。对于裁剪，我们应用了一种**不需要视觉示例**的简单裁剪方法。 C o u n t G D t x t 是仅在文本上训练的模型，与在示例和文本上训练的多模态 CountGD 不同，如表标题中指定的（粘贴在下面以供参考）。
> 
> <img alt="Image" width="836" src="https://private-user-images.githubusercontent.com/61301874/416615360-d7387453-38b0-466b-b081-b70549733e38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA0NzcxMTEsIm5iZiI6MTc0MDQ3NjgxMSwicGF0aCI6Ii82MTMwMTg3NC80MTY2MTUzNjAtZDczODc0NTMtMzhiMC00NjZiLWIwODEtYjcwNTQ5NzMzZTM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI1VDA5NDY1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiM2VhODYzNTM4MTdkOWQ0MTY5MzczNDM4ZmFiNGM0MTBkMGRlYTY2YmFhZWU1N2QyMGYxNTVhMTE1YzUwNTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DYMEdZ2Uu3cLFi0QmH88vS8ltiENvQwjR4RH3tnrMWE">一旦我有时间，我就可以发布纯文本版本。我们还发现，多模态 CountGD 仅在给定文本的情况下表现得相当好。它的效果可能不如仅使用文本训练的模型，但请尝试以下命令，看看您获得的性能如何。注意：'--simple_crop' 将图像分成 4 部分，并且只对所有裁剪中的计数求和，在任何阶段都不依赖示例。我还没有时间测试以下命令，所以如果您遇到任何错误，请告诉我，我会尽快回复。
> `python -u main_inference.py --output_dir ./countgd_val -c config/cfg_fsc147_val.py --eval --datasets config/datasets_fsc147_val.json --pretrain_model_path checkpoints/checkpoint_fsc147_best.pth --options text_encoder_type=checkpoints/bert-base-uncased --simple_crop --num_exemplars 0`
> 
> `python -u main_inference.py --output_dir ./countgd_test -c config/cfg_fsc147_test.py --eval --datasets config/datasets_fsc147_test.json --pretrain_model_path checkpoints/checkpoint_fsc147_best.pth --options text_encoder_type=checkpoints/bert-base-uncased --simple_crop --num_exemplars 0`

In the open source training code, I made the following modifications during the training phase to ensure that the visual exemplars are not involved in training, and obtain a model that was trained using only text was obtained：

`exemplars = [t["exemplars"][:0].to(device) for t in targets] `

Then, I using the following command to test：

`python -u main_inference.py --output_dir ./textonly_crop_test -c config/cfg_fsc147_test.py --eval --datasets config/datasets_fsc147_test.json --pretrain_model_path checkpoints/checkpoint_best_textonly_A6000.pth --options text_encoder_type=checkpoints/bert-base-uncased --num_exemplars 0`

And got **MAE=14.25, RMSE=129.45.**

I also tried using the simple cropping method in the testing phase, but I eventually realized that due to the incorrectly predict, the crop could not be triggered and finally get the same result as above（the predicted count is 15，less than args.num_select(900)):

![Image](https://github.com/user-attachments/assets/c99c1e47-0974-4bd0-818b-5b89268a554e)

So I changed the trigger for the crop：

`if args.simple_crop and pred_cnt == args.num_select:`
--------------->
`if args.simple_crop and gt_count > args.num_select:`

This triggers the crop correctly and gives a similar result to the paper：MAE=12.27, RMSE=98.81

But this trigger utilizes **gt_count** , which I don't think is appropriate in the testing phase.

Is it possible that I'm not training in the right way, and can you suggest a way to train my text-only model? Or is there any other option to trigger the drop appropriately?

thanks！！

**Comment by niki-amini-naieni** [comment]
Those errors are higher than what I have experienced, which means you could be training incorrectly for text only. Or, I may have used a lower confidence threshold for text only. Try grid searching the confidence thresholds of {0.15, 0.17, 0.19, 0.21, 0.23} to see which gives you the lowest validation MAE, and then use that. It could be the confidence threshold is too high. Right now, I am speculating on what could be the issue without going back to the code, since I have little time due to ICCV. Let's see if it is enough. Otherwise, I will have to return later.

**Comment by Eamon-0v0** [comment]
> confidence thresholds

Thank you! Wish you success in getting your paper accepted at ICCV!

I'm not in a hurry, so I'll try the suggestions you gave above, hope we can continue the discussion after your submission ends

Best wishes 😀

**Comment by niki-amini-naieni** [comment]
Thank you! Yes, I am happy to continue to discuss! Hopefully, we will be able to resolve all your questions (:

**Comment by Dodojordi** [comment]
Hi, I recently reproduced the results in the paper by adjusting the threshold conditions of the crop you had mentioned before, which is exciting. **Thank you very much for your suggestion, it is very useful to me.**🫡🫡

**Comment by niki-amini-naieni** [comment]
Great!

**Comment by Dodojordi** [comment]
Hi,I am very interested in how to use **only visual samples** for training and testing in ablation experiments. May I ask if you have removed the focal loss?🤔 We found that the algorithm cannot converge after removing the focal loss. Could you please provide a detailed introduction on how to train or provide training code? Thank you very much！

**Comment by niki-amini-naieni** [comment]
Hi @Dodojordi, I will see if I can find the training code. For a quick response for now, we used the focal loss. I will try to explain with an example. Suppose the training classes are ['apple', 'banana', 'fish', 'spider', 'tree']. Then, you would input the string 'apple. banana. fish. spider. tree.' to the model during training. For multi-modal CountGD, I inserted the exemplar right after the text associated with it. Let's say the exemplar is of a fish. Then, I would add the three exemplars where I put x (3 x's for 3 exemplars) in the string: 'apple. banana. fishxxx. spider. tree.' and apply self-attention between the exemplars of the fish and the text 'fish.' When using exemplars only, instead of removing the focal loss altogether, I just input the following: 'apple. banana. xxx. spider. tree.' and applied the classification loss as usual, so the queries corresponding to fish are pushed close the exemplar tokens and far away from the text tokens of 'apple,' 'banana,' 'spider,' and 'tree.'

**Comment by Dodojordi** [comment]
I think your explanation is clear enough. Thank you very much for your guidance！☺️

**Comment by niki-amini-naieni** [comment]
Great, no problem!
